### ISSUE TITLE
User feedback on slow navigation

### ISSUE THREAD

**Issue by akhenry** [nasa/openmct#5092]
**Is your feature request related to a problem? Please describe.**
Currently Open MCT gives no indication of activity when navigating between objects. This is not a problem on a fast connection, or when working locally, but it's an issue when the connection is slow. Users click on an object in the tree and it appears as though nothing at all is happening. The user needs immediate feedback on navigation, even if it takes a little while to actually load the requested object.

I believe that Open MCT used to navigate to a blank page until the object(s) had been loaded from the persistence store. This wasn't ideal, but at least gave the user immediate feedback when they navigated.

**Describe the solution you'd like**
A potential solution is some sort of activity indicator, much like what our telemetry tables do right now.

**Comment by charlesh88** [comment]
@akhenry This is a good callout and is something we should address. A big question: what can we update instantly, and what do we potentially need to wait to load? For example, can we instantly update the main view browse bar type icon and name and put a wait spinner (or an even cooler, simple CSS anim) in the object view area? 
![image](https://user-images.githubusercontent.com/1056412/164321238-73932418-5034-48e9-87c5-8c3307b253ac.png)
Does this vary per object type for any reason? Is there a delay loading the action menu items, or is that also instant?

**Comment by akhenry** [comment]
> what can we update instantly

This is a really good question. Technically speaking we have the object loaded already in the tree, so it's in memory. Our components are siloed though, so the tree and the object view are not sharing objects. In fact, I am now wondering whether this issue first arose when we removed object caching (which was a long time ago now).

So, in _some_ cases it would be possible to immediately update the name and type to the newly navigated object. I guess maybe there are cases where we are navigating via something other than the tree? But even in those cases I think we have always loaded the object once already.

TL;DR, I think we can immediately update the name and type with an API change.

One thing I want to avoid is flashing a spinner unnecessarily every time you navigate, even on fast connections, so I think I'd want to add a _slight_ delay to it?

**Comment by rukmini-bose** [comment]
I think the Charles' design with some sort of loading animation in the object view area would be effective. 

I don't know if I ever encountered this loading issue before so I have a question– when the requested object _does_ finally load, do the elements appear all at once or more one by one? Or better yet– is that something we would want to see? 

Also, how do we feel about users possibly different multiple objects due to the fact that an object may be taking quite some time to load? For example, they click on Object1, see it isn't loading, and try clicking on Object2 and/or Object3 in hope that it will load? I am not sure about how the backend works, so does this effect the performance in any way? 

For those reasons, another idea could be having an overlay over the entire page and having some sort of loading animation (example below). However, might prevent users from being able to click elsewhere and cancelling the object request. 

<img width="801" alt="Screen Shot 2022-04-20 at 4 58 33 PM" src="https://user-images.githubusercontent.com/48999852/164343512-0152d0de-07f2-4863-9ae1-76d32ae7e63b.png">

**Comment by charlesh88** [comment]
> One thing I want to avoid is flashing a spinner unnecessarily every time you navigate, even on fast connections, so I think I'd want to add a _slight_ delay to it?

Definitely agree. We can stub in a delay threshold before we show a loading element, and adjust it until it feels right. 

Our current wait spinner doesn't resize itself well. I'm going to make a new wait spinner that uses an SVG that will proportionally scale, so we can plop it into whatever context and have it be a size that's appropriate.

**Comment by charlesh88** [comment]
WIP at https://codepen.io/charlesh88/pen/zYpXWLV?editors=1100

**Comment by jvigliotta** [comment]
@rukmini-bose When a user selects an item on the the tree, if it takes a while and they select another item, that initial request is canceled on the backend. That's a good callout, because we implemented that a while ago to improve performance, previously those requests would still be going in the background. During that time, the discussion of preventing the user from clicking anything else had come up, but that was nixed as we don't want to restrict them.

**Comment by charlesh88** [comment]
@akhenry Could you have a look at https://codepen.io/charlesh88/pen/zYpXWLV?editors=1100 and let me know what you think? You did the earlier performance optimizations on the clock hands, and I've tried to utilize some of what was done there in addition to other performance research I've looked into.

**Comment by akhenry** [comment]
@charlesh88  I'm not so concerned with the performance of this load spinner because it should only appear temporarily, and never at the same time as telemetry. The problem with the clock hands is that they are omnipresent in real-time mode and they are taking CPU cycles away from rendering telemetry.

It looks good to me.

**Comment by akhenry** [comment]
@rukmini-bose Agree, we don't want to disable the tree for the reasons you highlighted. As @jvigliotta said we cancel load in this event by design, so we don't want to lose that.

So I _think_ we only want to overlay the object view.

**Comment by akhenry** [comment]
@charlesh88 @rukmini-bose Where are we at with this?
